### PR TITLE
Support for GSSAPI/Kerberos signed updates

### PR DESCRIPTION
**Pull request by bodgit**
This PR adds support for performing DNS updates using GSSAPI/Kerberos which makes it possible to update zones on Windows servers that have the zone require "Secure only" updates, (BIND can also be configured to support this mechanism).

To configure the provider, you need something like:
```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi = true
  }
}
```
Terraform needs to be run from a shell that has a valid Kerberos session which normally means having run `kinit` beforehand on Linux/Unix platforms. Current limitations are that it only supports using the credentials of the current shell running Terraform, you can't assume the identity of another user. This is mostly down to the fact I have yet to find a library for Go that both supports the necessary GSSAPI calls for doing the DNS exchange as well as the necessary Kerberos calls for creating a session as another user.

All of the heavy lifting is done by https://github.com/bodgit/tsig and it uses GSSAPI on Linux/Unix using https://github.com/apcera/gssapi via a native shared library or on Windows using SSPI through https://github.com/alexbrainman/sspi which has a very similar API to GSSAPI.

I've marked it WIP as it's currently reliant on a pending PR (miekg/dns#626) for the http://github.com/miekg/dns package to add support for non-HMAC TSIG algorithms but I figured I'd put the changes out there for people to test.

### CODE REVIEW

**Comment by apparentlymart** [comment]
Hi @bodgit! This looks interesting and useful... thanks for working on it.

Unfortunately it looks like the `gssapi` requires CGo, and we're not currently able to use CGo in the Terraform providers here since our release process uses cross-compilation and we don't have cross-toolkits available for other platforms in our build environment. :confounded:

Do you think it's possible to support this functionality using a pure-Go Kerberos implementation? I see [`gokrb5`](https://github.com/jcmturner/gokrb5) out there but I'm pretty rusty on Kerberos so I'm not sure if it supports sufficient functionality to get what we need to support Windows DNS updates here. I also realize that a re-implementation of the Kerberos functionality is non-ideal since the usual way to add Kerberos functionality to an application on Linux is via GSSAPI and so interop with other applications might not be perfect.

We've attempted to introduce CGo dependencies into providers before and found that it would require quite extensive updates to the release process which unfortunately we do not have the resources to do at this time due to other work. :disappointed:

**Comment by bodgit** [comment]
I wasn't sure if the CGo requirement was going to be an issue, makes sense.

I did look initially at using jcmturner/gokrb5 as it was a pure-Go implementation, (I have jcmturner/gokrb5#65 open from when I explored using it). From memory it works fine for the initial part of negotiating the context between the client and server however it currently (or did, I should check again) lacks an implementation of the `gss_get_mic()` and `gss_verify_mic()` functions which are used to create and verify the TSIG on messages which is why I then switched to the apcera/gssapi implementation.

However that implementation lacks a lot of the Kerberos-specific functions which then requires that you run `kinit` beforehand and basically set your session up before running Terraform, so switching would actually make this provider a bit easier to use in that regard; I just need to see how hard it is to get the missing functions implemented. It's one thing to use a Kerberos/GSSAPI library, quite another to implement one, but I'll see what I can do.

**Comment by bodgit** [comment]
Not sure how I closed this.

**Comment by arnisoph** [comment]
Any updates?

**Comment by bodgit** [comment]
@arnisoph I haven't had chance to revisit the pure-Go implementation, however the pivotal bit is the changes to the underlying DNS library; the main project maintainer seems to be ignoring the PR and insisting that I'm breaking backwards compatibility when I'm not and at least one other maintainer of the library seem to agree with me. I suspect my PR that is open there doesn't apply cleanly now however I'm in no rush to rebase it if it's just going to sit there and be ignored some more.

I'm not going to drop this as it's still actively being used, albeit just a custom in-house build off of this branch but the more people who can voice support would be useful.

**Comment by arnisoph** [comment]
We might need to maintain such a inhouse build, too. I'd like to prevent invoking the CLI tool nsupdate from TF..

**Comment by bodgit** [comment]
I've finally managed to improve the https://github.com/jcmturner/gokrb5 library with the missing bits so there is now a pure Go implementation which works. I just need the outstanding PR merged to the https://github.com/miekg/dns library.

**Comment by bodgit** [comment]
This is actually ready for review, and hopefully merge. The code now uses a pure Go Kerberos/GSSAPI implementation and is also therefore more versatile. It can be used as in my original example which means it relies on the current user context, or you can explicitly configure it with a realm, username and either a password, or a keytab (keytab not supported on Windows):
```terraform
provider "dns" {
  update {
    server   = "ns.example.com"
    gssapi   = true
    realm    = "EXAMPLE.COM"
    username = "user"
    password = "secret"
  }
}
```
or:
```terraform
provider "dns" {
  update {
    server   = "ns.example.com"
    gssapi   = true
    realm    = "EXAMPLE.COM"
    username = "user"
    keytab   = "/path/to/keytab"
  }
}
```
Now, the sticky bit is the maintainer of the underlying DNS library just ignored my PR and wouldn't engage in a discussion. To that end I've ended up embedding and extending enough of the library so there is a compatible `Exchange()` method with the same signature that understands the additional TSIG methods. It reuses as much of the library as possible but I've had to copy some non-exported functions.

**Comment by atkinson137** [comment]
@bodgit , is there any ETA on when this might be merged?

**Comment by bodgit** [comment]
@atkinson137 I need to fix up the handful of merge conflicts which have appeared in the meantime, however I don't have commit access to this repository so I'm/we're waiting on someone from Hashicorp to 👍 or 👎 this. 

I appreciate getting Terraform v0.12 ready has probably been taking up most of Hashicorp's time recently but there are a few issues and PR's that are just waiting for someone with commit privs. Sadly I don't think this provider gets as much love as say the AWS one, (happy to help with this!).

**Comment by bodgit** [comment]
Right, should be current against master again now.

**Comment by stephenpearson** [comment]
Would very much like to see this merged.  Hopefully this PR will get more attention soon.

File: website/docs/index.html.markdown [hunk 1]
```diff
@@ -31,13 +33,32 @@ resource "dns_a_record_set" "www" {
...
+Using GSS-TSIG (RFC 3645):
+
+```hcl
+provider "dns" {
+  update {
+    server   = "ns.example.com" # Using the hostname is important in order for an SPN to match
+    gssapi   = true
+    realm    = "EXAMPLE.COM"
+    username = "user"
+    keytab   = "/path/to/keytab"
```
Comment (apparentlymart) [inline website/docs/index.html.markdown]: From reading the configuration validation logic earlier, I got the sense that these `realm`, `username`, `password`, and `keytab` arguments are applicable only to `gssapi` mode.

If that is true, what do you think about reorganizing this a little so that these settings are inside a `gssapi` block, like this?:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {
      realm    = "EXAMPLE.COM"
      username = "user"
      keytab   = "/path/to/keytab"
    }
  }
}
```

We tend to use this configuration pattern when activating a feature implies requiring an additional set of arguments that would otherwise not be required. Arranging it this way would mean that the presence of the block activates the feature and all of the nested arguments could be marked as `Required: true`, and that may serve to simplify the validation logic by having Terraform and the provider SDK check these are present automatically.

In retrospect perhaps we ought to have originally put the existing key arguments in a `tsig` block too for symmetry, but I don't think that's important enough to warrant a breaking change now.

File: dns/provider.go [hunk 2]
```diff
@@ -90,6 +90,31 @@ func Provider() terraform.ResourceProvider {
 							Optional:    true,
 							DefaultFunc: schema.EnvDefaultFunc("DNS_UPDATE_KEYSECRET", nil),
 						},
+						"gssapi": &schema.Schema{
+							Type:        schema.TypeBool,
+							Optional:    true,
+							DefaultFunc: schema.EnvDefaultFunc("DNS_UPDATE_GSSAPI", nil),
```
Comment (apparentlymart) [inline dns/provider.go]: If the GSSAPI arguments are mutually-exclusive with the existing `key_` settings, we can tell the SDK that with an additional field here:

```go
    ConflictsWith: []string{"update.key_name", "update.key_algorithm", "update.key_secret"},
```

Specifying that will ask the SDK to validate that these are mutually exclusive and return an error to the user automatically if not.

File: website/docs/index.html.markdown (hunk 1 above)
Comment (bodgit) [reply website/docs/index.html.markdown]: Yes, I will do that. I think originally there was just a boolean toggle to enable GSSAPI as it picked everything else up from the environment due to the libraries used and then when I swapped the libraries to eliminate the CGo hurdle the parameters just organically grew from there. It should make the parameter validation much simpler, I agree it's a bit hairy now checking for the various allowed permutations.

File: dns/provider.go (hunk 2 above)
Comment (bodgit) [reply dns/provider.go]: That should make things much simpler.

**Comment by bodgit** [comment]
> Along with that, I'd like to also figure out how we might incorporate this into the existing test suite so that we can be more confident that we won't accidentally break it in future. In dns/acceptance.sh there is a small harness to run the provider's acceptance tests against two differently-configured instances of bind running in Docker containers. Do you know what it would take to add a third permutation of that which configures bind so it'll accept GSSAPI exchanges? I'm not sure if the container image it's using has a facility for activating that or if we'd need to adjust the contents of the image itself. 🤔

It will depend on the underlying OS the container is based on, I'm 99% sure the RHEL/CentOS packages have it enabled by default, for example. From memory it's two or three configuration options to point BIND at a keytab and tell it what principal to use and then there will be some tweaking of the zone ACL to grant permission to _Kerberized_ clients.

The tricky bit is you will also need a KDC container of some sort for the client to request a ticket from, as well as generate one or more keytab for BIND and/or the client, and that means the BIND container will also need to be configured with the address of the KDC. We might be able to get away with just statically creating everything once, or it will need to be regenerated every time the test suite runs.

File: website/docs/index.html.markdown (hunk 1 above)
Comment (bodgit) [reply website/docs/index.html.markdown]: @apparentlymart So the only edge case is that it is possible to enable GSSAPI without any further configuration for it to just use the current environment, so you would in theory end up with something like this:
```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {}
  }
}
```
Is that valid/acceptable? Alternatively, I could remove that ability so that the credentials are always explicitly configured in the provider.

Comment (apparentlymart) [reply website/docs/index.html.markdown]: This is a good question! Thanks.

Although it's currently not captured anywhere programmatically, we've noted before that in effect there are two different classes of provider configuration arguments: those that describe where things are to be created, and those that describe which human or automaton is running Terraform in terms that make sense to the remote system. While in most providers we _allow_ both to be set either in configuration or in the environment, we tend to recommend setting the "where"/"how" settings in the configuration but leaving the "who" settings for the environment or some other ambient source.

In this case, the `realm` argument feels to be to be roughly analogous to the AWS provider's `region` and `assume_role` arguments: they are more about where and how to create things than who is doing it. `assume_role` is, similar to `realm`, kinda on the edge of that: it is a _constraint_ on "who", but is still more like a "where"/"how".

With all of that theory out of the way, I'd expect that we'd want to show the primary way of using this with the `realm` set in the configuration but the other two settings coming from the environment:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
    gssapi {
      realm = "EXAMPLE.COM"
    }
  }
}
```

I read the above as "send updates to `ns.example.com`, which requires a credential from the `gssapi` realm `EXAMPLE.COM`". As a _user_, I then need to set up my environment to provide a suitable credential. The required realm is presumably chosen by the remote server and not by the user.

While on this subject, I wonder if it would make sense for the username and the keytab to be picked up from the logged in user by default, if possible. Kerberos and GSSAPI are normally used in "single sign-on" environments, where the user probably logged in to their current shell with Kerberos already and so their system username is probably a good default `username` and perhaps we can find a suitable `keytab` systematically somehow?

I'd still leave it so that all of these _can_ be configured in the environment, since it's pragmatic to give that flexibility for unusual situations, but I think we can design the configuration around a recommended "normal" pattern for what is and is not set, and document with that in mind.

How does all that seem to you? I assume you're working in a Kerberized environment already and so you'll have a more up-to-date sense of what we can and cannot determine automatically from the session; we usually try to make Terraform providers read from the "standard" credentials source for the system in question by default, rather than always requiring Terraform-specific configuration, but it's been so long since I was working in a Kerberos environment that I'm sure I'm forgetting details that are relevant here.

**Comment by apparentlymart** [comment]
The testing container image we're using right now is a third-party one, but it's seeming like we're getting into a specific enough scenario that we might be better off instead maintaining a Dockerfile in this repository so we can make tradeoffs that make testing easier even though they might not make sense for a production system, such as embedding a minimal KDC inside.

With that said, I know spinning up a whole new testing infrastructure is a lot of extra work to ask for when you've already put a lot of time into this, so I think a fine compromise for now would be for us to figure out together how one might _manually_ construct a suitable test environment using Docker -- ideally building on the image we're already using, but not necessarily -- and just document it as some steps for humans to run manually when working on Kerberos-related features, so at least future maintainers won't need to devise a whole testing strategy themselves.

If we were to think of this from the standpoint of starting up a container from the existing image and manually running some shell commands inside it, do you have a sense of what that might look like? No worries if not... I'm happy to help puzzle this out. From peeking at [the image's Dockerfile](https://github.com/drebes/docker-bind/blob/6ebd6c614f5e8171dc9e4ef4c6f231ed30f8e1b9/Dockerfile), it seems to be building on CentOS 7, so hopefully we could boil this down to running a few `yum` commands and copying in one or two example configuration files after spinning up a container from that image.

File: website/docs/index.html.markdown (hunk 1 above)
Comment (bodgit) [reply website/docs/index.html.markdown]: I think we've crossed wires slightly, however you've possibly answered the next question I had in mind.

The use case I was talking about was using the currently logged in user, (probably shouldn't have said environment in that case!), if you don't have any configuration (apart from enabling GSSAPI support) _and_ none of the GSSAPI-specific provider environment variables are set then it will try and use a ticket for the current user, (this was initially the only way it would work due to the first GSSAPI library I used having no Kerberos smarts). No other configuration is then needed as it will pick up everything automatically, you just need a way to say "do GSSAPI" hence conceptually you could end up with just an empty `gssapi {}` block in the provider as you still need to enable it.

A keytab can be utilised automatically in the above scenario, provided it has been provided to `kinit` previously. Something like `kinit -F -k -t /path/to/keytab username` IIRC. According to https://web.mit.edu/kerberos/krb5-devel/doc/basic/keytab_def.html#default-client-keytab there are some Kerberos-level defaults that can be tried as well to locate a keytab.

My next question I thought of is that currently the provider allows environment variables for literally every setting, and I was wondering if that is desired or not? Also, the code seems to duplicate the environment variable way of retrieving settings, i.e.:
```golang
"server": {
        Type:        schema.TypeString,
        Required:    true,
        DefaultFunc: schema.EnvDefaultFunc("DNS_UPDATE_SERVER", nil),
},
```
and then further down in `configureProvider()`:
```golang
if len(os.Getenv("DNS_UPDATE_SERVER")) > 0 {
        server = os.Getenv("DNS_UPDATE_SERVER")
} else {
        return nil, nil
}
```
Is this correct?

Also, in the context of using the current user, there would be no way of expressing "do GSSAPI" without some sort of flag given there won't be any other GSSAPI provider environment variables set. So maybe the provider could be made to try GSSAPI if there are no explicit TSIG settings and then fall back to unsigned updates, (i.e. no TSIG of any sort)? What do you think?

**Comment by bodgit** [comment]
> If we were to think of this from the standpoint of starting up a container from the existing image and manually running some shell commands inside it, do you have a sense of what that might look like? No worries if not... I'm happy to help puzzle this out. From peeking at the image's Dockerfile, it seems to be building on CentOS 7, so hopefully we could boil this down to running a few yum commands and copying in one or two example configuration files after spinning up a container from that image.

If the current BIND container is based on CentOS 7 then it should definitely have the GSSAPI support compiled in. I'm not sure if putting the KDC on the same host makes things easier or not though. You might also want to run the tests from a separate client container to prevent conflicting with any existing Kerberos settings on the underlying host.

Off the top of my head, assuming building it from scratch every time, I think the general flow would be:

1) Create KDC container, configure Kerberos, create database `kdb5_util create ...`
2) Create BIND container, configure Kerberos, add principal `kadmin addprinc ...`, add keytab `kadmin ktadd ...`
3) Create client container, configure Kerberos, add principal and keytab
4) Test Terraform provider on client pointing to BIND container

There are numerous guides that all cover the Kerberos basics, i.e. https://major.io/2012/02/05/the-kerberos-haters-guide-to-installing-kerberos/ , https://www.howtoforge.com/how-to-setup-kerberos-server-and-client-on-ubuntu-1804-lts/ , etc.

My environment is a Windows AD domain so the KDC side is all hidden from me, I've only really had to deal with the client side which is either Linux or Windows.

File: website/docs/index.html.markdown (hunk 1 above)
Comment (apparentlymart) [reply website/docs/index.html.markdown]: Looking at this again today with fresh eyes, I see that my initial thought of `realm` effectively being required isn't really true, because in real-world Kerberos environments there can often be multiple realms that are all equally fine to use, separated only for administrative purposes, and so it would be reasonable to have a DNS server that accepts updates from multiple different realms and have the end-user running Terraform be the one to decide. And given that, having it pick up an "ambient" realm for the logged in user makes sense, reducing the configuration to just an empty `gssapi {}` block. An empty block is a little unconventional, but not without precedent and I think worthwhile given the other benefits it gives us in clarifying which settings are GSSAPI-specific.

If there's a reasonable way to reduce the required configuration burden down to _just_ the server in the already-logged-in-with-GSSAPI scenario then I think that would be awesome:

```hcl
provider "dns" {
  update {
    server = "ns.example.com"
  }
}
```

This is consistent with the goal of having the configuration describe _where_ while the _who_ comes from ambient information, and I think that's the ideal case. Since there isn't (as far as I know) a "standard" way to set the TSIG settings ambiently we invented a provider-specific environment variable scheme for those, but I'd always prefer a mechanism that re-uses existing conventions where it's available, so having the provider try to use GSSAPI by default if the TSIG settings are absent seems great to me.

Regarding environment variables for every setting: that tends to be the case in most providers mainly because it makes test automation more convenient. In practice I think we'd expect end-users _not_ to use the `DNS_UPDATE_SERVER` environment variable, but we tend to be pragmatic about these things and rely on recommendations and good examples rather than forcing a particular usage pattern.

**Comment by apparentlymart** [comment]
Cool... thanks for those extra details. It is seeming like our existing testing strategy isn't going to scale well here and so we'll need something a little more sophisticated, but if I can get to the point where I can manually construct an environment to test this PR without affecting my main system configuration then I can write those instructions down as a first pass, and we can then figure out how we might automate such tests separately.

I believe for this provider we're currently running the acceptance tests as part of the Travis-CI run, so that environment likely imposes some additional constraints on us that we'd need to consider in order to fully automate this, and so I'd rather get a "good enough" semi-manual solution in place than to block on totally reworking the entire test procedures here.

With that said, I'm going to wait until the configuration bits are finalized before I try to do this since I want to test it in its final state. As you mentioned, it'd probably be best for the client to be dockerized too so that it's not dependent on any global client configuration on the development host but yet we can still pick up the necessary credentials ambiently.

**Comment by bodgit** [comment]
@apparentlymart I've refactored the configuration so it now has a dedicated `gssapi {}` block which seems to work however I'm struggling to get the `ConflictsWith` settings to work. I'm wondering if this is down to the `update {}` block being configured as a `TypeList` and I'm hitting hashicorp/terraform#11101 or similar? Should these be a `TypeMap` or `TypeSet` instead?

**Comment by apparentlymart** [comment]
Thanks for these updates, @bodgit!

Indeed I think you're right that `ConflictsWith` won't work here, because it doesn't work for collection types. I think we may need to hand-write some validation logic in the `configureProvider` function after all, but at least now it's hopefully a little more straightforward in that it only has to deal with checking the presence of each of the blocks, and not with the individual settings inside.

**Comment by atkinson137** [comment]
@bodgit Anyway I can help out with this?

**Comment by sean-nixon** [comment]
@apparentlymart @bodgit I was curious if there were any plans to revisit this relatively old PR? The organization I'm working with is very much interested in the ability to manage secure DNS updates for AD-integrated DNS.

**Comment by bodgit** [comment]
For anyone still following this PR. I see this provide has been refactored slightly to use an SDK so I will fix up this PR so it's applying cleanly again but ideally I need support from HashiCorp to get this over the line.

**Comment by bhoriuchi** [comment]
@bodgit if you get the chance to investigate. We have been using a slightly modified version of your fork with some changes to dynamically create the krb5 config in memory vs. using a file and everything has worked well except for wildcard dns entries in windows DNS. for whatever reason they seem to fail with what seems to be auth errors.

**Comment by bodgit** [comment]
I've not forgotten about this PR; I've actually had another go at getting the upstream DNS library to accept a PR to allow plugging in alternative TSIG mechanisms. This would mean I don't need to carry around a forked copy of the DNS client. One maintainer has approved, I'm just waiting hopefully on a merge.

I'm also investigating trying to spin up some Kerberos test infrastructure as part of the acceptance testing so we can prove it works. Just trying to work out how to spin up the various Docker containers and get keytab files generated and distributed.

@bhoriuchi I think I've noticed the same issue with trying to add wildcard entries, I'll try and investigate and see what's wrong. It shouldn't be specific to GSS-TSIG as it just hashes the same stripped DNS message, it shouldn't care that there's a wildcard.

**Comment by apparentlymart** [comment]
Hi all! I'm sorry this PR ended up being a multi-year epic. :confounded: 

Unfortunately I can't do much to change that situation now since in the intervening time folks and teams have reorganized and I'm no longer one of the maintainers of this repository. I think @bodgit's assessment earlier that most of HashiCorp's provider development energy ends up going into the "major cloud" providers is accurate, and so to be honest I'm not sure if anyone is proactively monitoring this repository for contributions anymore, since it's not the most commonly-used of Terraform providers.

I don't want the work here to be lost, but I also can't _directly_ help with getting it merged beyond the code review I already did here. It's tough to include a change this substantial in a provider that's otherwise essentially in "maintenance mode" only, and thus there isn't necessarily anyone with sufficient context to review and support it.

One possible alternative path forward would be to exploit the fact that Terraform v0.13 and later can install providers published by parties other than HashiCorp. If someone were to fork this repository into their own account and merge this PR there then they could potentially publish their own version of this provider in their own namespace on [Terraform Registry's provider index](https://registry.terraform.io/browse/providers), as a "Community" provider. Indeed, it seems like two people have already published their own versions of this provider in the way I'm describing:

* [`tidalf/dns`](https://registry.terraform.io/providers/tidalf/dns/latest/docs)
* [`someara/dns`](https://registry.terraform.io/providers/someara/dns/latest/docs)

I can't say what is or is not included in those or to what degree these folks are supporting these for use by others, but I share this just to illustrate what I mean, not to recommend these or to suggest that they might have `gssapi` support already. (I think they don't, based on the documentation.)

**Comment by bodgit** [comment]
My upstream DNS changes have been merged, literally an hour after my previous comment, so this PR should become a lot simpler. Most of the churn is annoyingly in the vendored modules which I think is how this PR earned the XXL label.

I'm going to raise a couple of things as separate PR's to make this one hopefully smaller:
* Bump the github.com/miekg/dns dependency to the latest version that has my API changes.
* Tweak the acceptance testing setup. I've worked out how to get Kerberos KDC and BIND instances running in Docker so that should give us better confidence in this PR.

**Comment by bodgit** [comment]
I've now raised two separate PR's to update the DNS library dependency and also change the acceptance test infrastructure so that Kerberos signed updates can now be verified.

**Comment by bodgit** [comment]
Ok, this is now rebased against #118 and #119. I've kept the commits logically separate so there's one commit that adds my GSS-TSIG library and its dependences to the vendored packages, another that updates the code and one more that updates the acceptance tests.

There are now tests configured that run against the Kerberos-enabled BIND name server. I use the same test principal but configured three different ways; username & password, username & keytab, and finally relying on an existing ticket/session. They seem to be passing 🎉 🍾 

Based on the suggestion from @apparentlymart the provider is configured with the following:
```terraform
provider "dns" {
  server = "ns.example.com"
  gssapi {
    realm    = "EXAMPLE.COM" # Required
    username = "test" # Optional but then requires one of the following as well
    password = "password"
    keytab   = "/path/to/keytab"
  }
}
```
It also uses the `$DNS_UPDATE_REALM`, `$DNS_UPDATE_USERNAME`, `$DNS_UPDATE_PASSWORD`, and `$DNS_UPDATE_KEYTAB` environment variables.

**Comment by bodgit** [comment]
@kmoe @apparentlymart any chance I could possibly get a review on this and the other two PR's please?

They look scary but that's purely down to touching the `vendor` directory, which is why I've kept them as separate PR's or commits. Incidentally, I notice some TF providers (AWS for instance) have now dropped vendoring modules.

**Comment by bhoriuchi** [comment]
@bodgit just wondering if you were able to address the wildcard DNS issue with the latest changes? Additionally wondering if the provider still requires a krb5.conf file or if there is opportunity to load the config from a variable? Either way thanks for keeping this effort going!

**Comment by bodgit** [comment]
@bhoriuchi It appears Windows doesn't support creating wildcard DNS entries using dynamic DNS updates. I've verified using a small piece of test code to send the wildcard update; it doesn't work against Windows but works fine against a Kerberos-enabled BIND. If I create a non-wildcard record it succeeds on both. I've also tried using `nsupdate` to send the update and that doesn't work either so I'm reasonably confident it's a bug/implementation detail on the Windows side, which is annoying.

My code (hopefully) honours all of the standard Kerberos environment variables so it will use the `$KRB5_CONFIG` environment variable if it's set. As for passing the file contents in as a variable, that isn't really portable to Windows as it has no relevance there so it needs to be done in a way that keeps the API sane. That then also needs to be surfaced by this provider code, does that just have a parameter with a raw krb5.conf as heredoc? I have some ideas on how to do it in a clean way but I'd like some feedback on this PR first before I change it any further. It can always be an incremental update later.

**Comment by kmoe** [comment]
@bodgit Thank you for your saintly patience and the work you've put in to this PR. After some internal reshuffling at HashiCorp, this repository is now maintained by me. Since my team is responsible for more than 20 other repositories, this maintenance will be rather slow, especially for new features, since, as Martin indicates, our top priority is the stability and correctness of these very widely used providers. They are being monitoried for contributions, however, and priority is based on a number of factors including popularity measured in :+1: reactions.

As mentioned in https://github.com/hashicorp/terraform-provider-dns/pull/118, I'd like to avoid a major version bump of this provider, so it's best if we continue supporting the (admittedly deprecated) HMAC-MD5 algorithm. Once that is added I think we can merge all three PRs at the same time.

**Comment by bodgit** [comment]
@kmoe I knew there was something I'd forgotten to mention in my comments, HMAC-MD5 should be supported still. The PR to the upstream DNS library I had merged allows you to plug in your own TSIG implementations so I just provide my own that still supports HMAC-MD5 as well as all of the other algorithms.

I've added a test case to make sure it works however I'm having trouble getting Travis to actually run the tests; it looks like it has such a backlog of jobs to run that the tests for this repo never manage to make it to the front of the queue before they get cancelled after something like 5 or 6 hours. I've checked this morning and the backlog is still over 2,600 jobs with only ~ 240 running at any one time. I will periodically keep trying to schedule this and I'll squash the empty commits away once I get things to run.

**Comment by kmoe** [comment]
@bodgit Thanks for adding the MD5 test. There's an issue with Travis at the moment that I'll be investigating today. Unfortunate timing, as it was working last week. By the end of today, either we'll have Travis working again, or we'll start migrating onto our new CI system and I'll approve this PR after the tests pass locally.

**Review by kmoe — APPROVED** [review]
We can merge this once the GitHub Actions setup is complete. Certain acceptance tests aren't working yet, for reasons unrelated to this PR. Thanks for your patience!

**Comment by bodgit** [comment]
If you need me to look at anything or rebase things, let me know.

**Comment by kmoe** [comment]
@bodgit I'd like to merge this soon despite the ongoing GitHub Actions problem (which is unrelated to this PR). Tests were all passing locally when I last checked, but I ran them again just now and get the following error:

```
=== RUN   TestAccDnsTXTRecordSet_Basic
    resource_dns_txt_record_set_test.go:41: Step 1/6 error: Error running apply: 
        Error: Error updating DNS record: Error negotiating GSS context: dial tcp 92.242.132.24:53: i/o timeout
```

Would you mind rebasing and taking a look at this? If I have time later today I'll investigate myself.

**Comment by bodgit** [comment]
Sure. Where's that external DNS server coming from though? If I do a reverse lookup it comes back as `unallocated.barefruit.co.uk`, is that a personal domain?

**Comment by kmoe** [comment]
Bizarre, I have no idea. All I did was pull your branch locally and run `./internal/provider/acceptance.sh`.

**Comment by kmoe** [comment]
Ha, TIL! https://www.barefruit.com/

> Barefruit generates highly targeted traffic for ISPs by replacing DNS and HTTP errors with relevant advertising.

Well then...

**Comment by kmoe** [comment]
With 8.8.8.8:
```       
--- FAIL: TestAccDnsSRVRecordSet_Basic (0.75s)
=== RUN   TestAccDnsTXTRecordSet_Basic
    resource_dns_txt_record_set_test.go:41: Step 1/6 error: Error running apply: 
        Error: Error updating DNS record: Error negotiating GSS context: dial tcp: lookup ns.example.com: no such host
```
Not sure why I'm getting this. Do the tests pass for you, @bodgit ?

**Comment by bodgit** [comment]
If you're running them locally, you might have to pop a `127.0.0.1 ns.example.com` entry into your `/etc/hosts` (assuming you're on a Unix-ish system, there's an equivalent on Windows buried under `C:\windows\` somewhere). I think I did that step as part of the Travis job setup.

**Comment by bodgit** [comment]
There's something screwy in my install, if I try and run the tests on `main`, or this branch, it always errors out here:
```
+ make testacc TEST=./internal/provider
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/provider -v  -timeout 120m
=== RUN   TestAccDataDnsARecordSet_Basic
--- PASS: TestAccDataDnsARecordSet_Basic (20.96s)
=== RUN   TestAccDataDnsAAAARecordSet_Basic
--- PASS: TestAccDataDnsAAAARecordSet_Basic (10.23s)
=== RUN   TestAccDataDnsCnameRecordSet_Basic
--- PASS: TestAccDataDnsCnameRecordSet_Basic (10.32s)
=== RUN   TestAccDataDnsMXRecordSet_Basic
--- PASS: TestAccDataDnsMXRecordSet_Basic (10.35s)
=== RUN   TestAccDataDnsNSRecordSet_Basic
--- PASS: TestAccDataDnsNSRecordSet_Basic (10.41s)
=== RUN   TestAccDataDnsPtrRecordSet_Basic
--- PASS: TestAccDataDnsPtrRecordSet_Basic (10.15s)
=== RUN   TestAccDataDnsSRVRecordSet_Basic
--- PASS: TestAccDataDnsSRVRecordSet_Basic (10.45s)
=== RUN   TestAccDataDnsTxtRecordSet_Basic
--- PASS: TestAccDataDnsTxtRecordSet_Basic (15.23s)
=== RUN   TestHashIPString
--- PASS: TestHashIPString (0.00s)
=== RUN   TestProvider
--- PASS: TestProvider (0.00s)
=== RUN   TestProvider_impl
--- PASS: TestProvider_impl (0.00s)
=== RUN   TestAccDnsARecordSet_Basic
--- FAIL: TestAccDnsARecordSet_Basic (3.49s)
panic: interface conversion: interface {} is nil, not *provider.DNSClient [recovered]
	panic: interface conversion: interface {} is nil, not *provider.DNSClient

```

**Comment by kmoe** [comment]
Thank you, I had entirely neglected the `127.0.0.1 ns.example.com` entry. This may well be what's broken in the GHA setup as well - will check later. Tests all passing for me now on both this branch and `main`.

The error you're getting seems to indicate you've somehow ended up with a nil client, despite (presumably) setting `DNS_UPDATE_SERVER`. I don't presently see how this is possible without the provider erroring out during `New()`. Are you running the tests via `acceptance.sh`? Only thing I can think of at the moment is to try in a clean env. Worth double-checking you don't have any `terraform.*` or `.terraform*` files lying around also. 

Were the tests passing for you previously?

**Comment by bodgit** [comment]
I think my error was caused by me having a 0.12.x `terraform` binary somewhere in my path which was being used for running the tests. I don't think the `configureProvider()` function was being called, which is what goes and looks up all of the environment variables to configure the provider. Once I removed that binary, the test suite seemed to be downloading a fresh copy of `terraform` for each individual test, which is really slow on my internet connection, but the tests passed. I've now installed a 0.14.x `terraform` binary and tests are still passing and running a bit quicker. All good here now.

Not having the `/etc/hosts` record makes sense now with the first DNS provider you were using as they have that annoying habit of serving a landing page address for any `NXDOMAIN` response so you were looking up `ns.example.com`, getting a landing page address and then trying to negotiate a Kerberos ticket with it! I hate those sort of sites.

**Comment by kmoe** [comment]
Fabulous. After three years and a number of related and unrelated difficulties along the way, I'm now going to merge this PR. The new functionality will go out in today's v3.1.0 release. Thank you @bodgit for your great work on this PR and @apparentlymart for the thorough review.

**Comment by bodgit** [comment]
198601-She-Went-A-Little-Too-Far
